File: leo2-node/src/main/java/org/deku/leo2/node/rest/ExceptionMapper.java
package org.deku.leo2.node.rest;

import com.fasterxml.jackson.core.JsonLocation;
import com.fasterxml.jackson.core.JsonParseException;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonMappingException;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import sx.util.Cast;

import javax.inject.Named;
import javax.ws.rs.WebApplicationException;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.ext.Provider;
import java.util.Optional;

/**
 * Created by masc on 21.04.15.
 */
@Named
@Provider
public class ExceptionMapper implements javax.ws.rs.ext.ExceptionMapper<Exception> {
    /**
     * Represents webservice result on exception
     */
    static class ExceptionResult {
        private int mStatus;
        private String mMessage;

        public ExceptionResult(int status, String message) {
            mStatus = status;
            mMessage = message;
        }

        public ExceptionResult(int status, Exception e) {
            this(status, e.getMessage());
        }

        public Integer getStatus() {
            return mStatus;
        }

        public String getMessage() {
            return mMessage;
        }
    }

    Log mLogger = LogFactory.getLog(ExceptionMapper.class);

    @Override
    public javax.ws.rs.core.Response toResponse(Exception e) {
        ExceptionResult result;
        if (e instanceof WebApplicationException) {
            WebApplicationException we = (WebApplicationException) e;
            result = new ExceptionResult(we.getResponse().getStatus(), we);
        } else if (e instanceof JsonMappingException) {
            JsonMappingException jm = (JsonMappingException) e;
            String locationMessage = String.join(".", (Iterable)jm.getPath().stream().map(p -> p.getFieldName())::iterator);

            result = new ExceptionResult(Response.Status.NOT_FOUND.getStatusCode(),
                    String.format("JSON mapping error [%s]: %s", locationMessage, jm.getCause().getMessage()));
        } else if (e instanceof JsonProcessingException) {
            JsonParseException je = (JsonParseException) e;
            JsonLocation jl = je.getLocation();

            Optional<String> locationMessage = Optional.empty();
            if (jl != null) {
                locationMessage = Optional.of(String.format("in line %d column %d", jl.getLineNr(), jl.getColumnNr()));
            }

            result = new ExceptionResult(Response.Status.NOT_FOUND.getStatusCode(),
                    String.format("JSON parse error%s: %s",
                            locationMessage.orElse(""),
                            je.getOriginalMessage()));
        } else {
            result = new ExceptionResult(Response.Status.NOT_FOUND.getStatusCode(), e);
        }

        return Response
                .status(result.getStatus())
                .entity(result)
                .type(MediaType.APPLICATION_JSON_TYPE)
                .build();
    }
}

File: leoz-central/src/main/java/org/deku/leoz/central/rest/ExceptionMapper.kt
package org.deku.leoz.central.rest

import org.deku.leoz.central.Application
import org.deku.leoz.central.data.isJooqAccessException
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Component
import org.zalando.problem.Status
import sx.rs.RestProblem
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response
import javax.ws.rs.ext.Provider

/**
 * Created by masc on 21.04.15.
 */
@Profile(Application.PROFILE_CENTRAL)
@Component
@Provider
class ExceptionMapper : org.deku.leoz.node.rest.ExceptionMapper() {
    private val log = LoggerFactory.getLogger(this.javaClass)

    override fun toResponse(e: Exception): javax.ws.rs.core.Response {
        val status: Int
        val entity: Any?

        when {
            e.isJooqAccessException() -> {
                log.error(e.message)
                status = Response.Status.INTERNAL_SERVER_ERROR.statusCode
                entity = RestProblem(status = Status.INTERNAL_SERVER_ERROR, detail = e.message)
            }
            else -> {
                return super.toResponse(e)
            }
        }

        return Response
                .status(status)
                .entity(entity)
                .type(MediaType.APPLICATION_JSON_TYPE)
                .build()
    }
}